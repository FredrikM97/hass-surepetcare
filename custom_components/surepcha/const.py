""" "Constants for the Sure Petcare integration."""

DOMAIN = "surepcha"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
ENTRY_ID = "entry_id"
SCAN_INTERVAL = 300
POLLING_SPEED = "polling_speed"
LOCATION_INSIDE = "location_inside"
LOCATION_OUTSIDE = "location_outside"
OPTION_DEVICES = "devices"
DEVICE_OPTION = "device_option"
OPTIONS_FINISHED = "finished"
PRODUCT_ID = "product_id"
DEVICES = "devices"
NAME = "name"
