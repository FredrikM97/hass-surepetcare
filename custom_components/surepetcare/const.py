"""TODO."""

DOMAIN = "surepetcare"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
