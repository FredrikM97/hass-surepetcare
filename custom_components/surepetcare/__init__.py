"""TODO."""

import logging
from typing import Any, List

from surepetcare.client import SurePetcareClient
from surepetcare.household import Household
from surepetcare.enums import ProductId

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
from homeassistant.core import Event, HomeAssistant, callback
from homeassistant.helpers import device_registry as dr

from .const import (
    CLIENT_DEVICE_ID,
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    FACTORY,
    KEY_API,
    TOKEN,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator

logger = logging.getLogger(__name__)

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> bool:
    """TODO."""
    logger.info("async_setup_entry called for entry_id=%s", entry.entry_id)

    surepetcare_data: dict[str, Any] = {}
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = surepetcare_data

    client: SurePetcareClient = SurePetcareClient()
    # Try login with token+device_id, else fallback to email+password
    token = entry.data.get(TOKEN)
    device_id = entry.data.get(CLIENT_DEVICE_ID)
    email = entry.data.get("email")
    password = entry.data.get("password")
    if token and device_id:
        await client.login(token=token, device_id=device_id)
    elif email and password:
        await client.login(email=email, password=password)
    else:
        raise Exception("No valid credentials found in config entry data")
    surepetcare_data[FACTORY] = client

    async def on_hass_stop(event: Event) -> None:
        """Close connection when hass stops."""
        await client.close()

    # Setup listeners
    entry.async_on_unload(
        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
    )
    try:
        households: List[Household] = await client.api(Household.get_households())
        entities = []
        for household in households:
            entities.extend(await client.api(household.get_pets()))
            entities.extend(await client.api(household.get_devices()))
            # Add each hub device to hubs list
        await client.close()
    except Exception as exc:
        await client.close()
        raise Exception("Configuration not finished") from exc

    remove_stale_devices(hass, entry, entities)
    # Setup the device coordinators
    coordinator_data = {
        KEY_API: client,
        COORDINATOR_DICT: {},
    }

    for device in entities:
        if device.product != ProductId.HUB:
            continue
        entities.remove(device)
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    for device in entities:
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    surepetcare_data[COORDINATOR] = coordinator_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        surepetcare_data = hass.data[DOMAIN].pop(entry.entry_id)
        client: SurePetcareClient = surepetcare_data[FACTORY]
        await client.close()

    return unload_ok


@callback
def remove_stale_devices(
    hass: HomeAssistant, config_entry: ConfigEntry, devices: list[Any]
) -> None:
    """Remove stale devices from device registry. TODO: Work in progress and not functional yet"""

    device_registry = dr.async_get(hass)
    device_entries = dr.async_entries_for_config_entry(
        device_registry, config_entry.entry_id
    )
    all_device_ids = {str(device.id) for device in devices}
    for device_entry in device_entries:
        device_id: str | None = None  # Only define here
        # Check that device part of DOMAIN
        for identifier in device_entry.identifiers:
            if identifier[0] != DOMAIN:
                continue

            _id = identifier[1]
            device_id = str(_id)
        if device_id is None or device_id not in all_device_ids:
            logger.info(
                "Removing stale device entry %s for config entry %s",
                device_entry.id,
                config_entry.entry_id,
            )
            device_registry.async_update_device(
                device_entry.id, remove_config_entry_id=config_entry.entry_id
            )
