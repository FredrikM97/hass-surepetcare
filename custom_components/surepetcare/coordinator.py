from datetime import timedelta
import logging
from typing import Any

from surepetcare.client import SurePetcareClient
from surepetcare.devices.device import SurepyDevice

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

logger = logging.getLogger(__name__)

SCAN_INTERVAL = 300


class SurePetCareDeviceDataUpdateCoordinator(DataUpdateCoordinator[Any]):
    """Coordinator to manage data for a specific SurePetCare device."""

    config_entry: ConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
        config_entry: ConfigEntry,
        client: SurePetcareClient,
        device: SurepyDevice,
    ) -> None:
        """Initialize device coordinator."""
        self.client = client
        self.device = device
        self._exception: Exception | None = None

        super().__init__(
            hass,
            logger,
            config_entry=config_entry,
            name=f"Update coordinator for {device}",
            update_interval=timedelta(seconds=SCAN_INTERVAL),
        )

    @callback
    def _observe_update(self, device: Any) -> None:
        """Update the coordinator for a device when a change is detected."""
        self.async_set_updated_data(data=device)

    async def _async_update_data(self) -> Any:
        """Fetch data from the api for a specific device."""
        return await self.client.api(self.device.refresh())
