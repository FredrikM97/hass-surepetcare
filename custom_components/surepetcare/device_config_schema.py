"""TODO."""

from voluptuous import Required

from surepetcare.enums import ProductId

DEVICE_CONFIG_SCHEMAS = {
    ProductId.FEEDER_CONNECT: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.HUB: {
        "schema": None,
    },
    ProductId.PET: {
        "schema": None,
    },
}
