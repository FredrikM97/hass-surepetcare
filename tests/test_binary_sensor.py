from unittest.mock import MagicMock, patch
import pytest
from custom_components.surepetcare import binary_sensor
from custom_components.surepetcare.binary_sensor import SurePetCareSensor, SENSORS, ProductId
from tests.conftest import DummyDevice, DummyCoordinator, DummyClient


def test_binary_sensor_entity_refresh():
    desc = SENSORS[ProductId.FEEDER_CONNECT][0]
    c = DummyCoordinator(DummyDevice())
    s = SurePetCareSensor(c, DummyClient(), desc)
    s._refresh()
    assert s._attr_native_value is not None

@pytest.mark.asyncio
async def test_async_setup_entry_binary_sensor():
    hass = MagicMock()
    config_entry = MagicMock()
    async_add_entities = MagicMock()
    device = DummyDevice()
    coordinator = DummyCoordinator(device)
    coordinator.device.product_id = ProductId.FEEDER_CONNECT
    from custom_components.surepetcare.const import KEY_API, COORDINATOR_LIST
    hass.data = {"surepetcare": {config_entry.entry_id: {"coordinator": {KEY_API: DummyClient(), COORDINATOR_LIST: [coordinator]}}}}
    config_entry.subentries = {"1": MagicMock(data={"id": "1"})}
    with patch("custom_components.surepetcare.binary_sensor.SENSORS", {ProductId.FEEDER_CONNECT: SENSORS[ProductId.FEEDER_CONNECT]}):
        await binary_sensor.async_setup_entry(hass, config_entry, async_add_entities)
    async_add_entities.assert_called()
