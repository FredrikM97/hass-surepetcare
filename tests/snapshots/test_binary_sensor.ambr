# serializer version: 1
# name: test_platform_setup_and_discovery[binary_sensor.hem_hub_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.hem_hub_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '295972-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.hem_hub_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Hem-hub Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.hem_hub_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.katzenklappe_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '1299453-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Katzenklappe Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.katzenklappe_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '1299453-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Katzenklappe Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_none-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.katzenklappe_none',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'curfew',
    'unique_id': '1299453-curfew',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.katzenklappe_none-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'control.curfew': list([
        dict({
          'enabled': False,
          'lock_time': datetime.time(22, 0),
          'unlock_time': datetime.time(22, 1),
        }),
        dict({
          'enabled': False,
          'lock_time': datetime.time(22, 6),
          'unlock_time': datetime.time(23, 6),
        }),
      ]),
      'friendly_name': 'Katzenklappe None',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_none',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.laundry_door_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.laundry_door_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '727608-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.laundry_door_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Laundry Door Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.laundry_door_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.laundry_door_none-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.laundry_door_none',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'curfew',
    'unique_id': '727608-curfew',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.maui_matlada_old_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.maui_matlada_old_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '269654-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.maui_matlada_old_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Maui matl√•da (old)  Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.maui_matlada_old_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.maui_matlada_old_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.maui_matlada_old_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '269654-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery[binary_sensor.maui_matlada_old_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Maui matl√•da (old)  Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.maui_matlada_old_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.hem_hub_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.hem_hub_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '295972-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.hem_hub_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Hem-hub Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.hem_hub_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.katzenklappe_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '1299453-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Katzenklappe Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.katzenklappe_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '1299453-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Katzenklappe Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_none-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.katzenklappe_none',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'curfew',
    'unique_id': '1299453-curfew',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.katzenklappe_none-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'control.curfew': list([
        dict({
          'enabled': False,
          'lock_time': datetime.time(22, 0),
          'unlock_time': datetime.time(22, 1),
        }),
        dict({
          'enabled': False,
          'lock_time': datetime.time(22, 6),
          'unlock_time': datetime.time(23, 6),
        }),
      ]),
      'friendly_name': 'Katzenklappe None',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.katzenklappe_none',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.laundry_door_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.laundry_door_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '727608-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.laundry_door_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Laundry Door Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.laundry_door_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.laundry_door_none-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.laundry_door_none',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'curfew',
    'unique_id': '727608-curfew',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.maui_matlada_old_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.maui_matlada_old_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'connectivity',
    'unique_id': '269654-connectivity',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.maui_matlada_old_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Maui matl√•da (old)  Connectivity',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.maui_matlada_old_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.maui_matlada_old_learn_mode-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.maui_matlada_old_learn_mode',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Learn Mode',
    'platform': 'surepetcare',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'learn_mode',
    'unique_id': '269654-learn_mode',
    'unit_of_measurement': None,
  })
# ---
# name: test_platform_setup_and_discovery_missing_entities[binary_sensor.maui_matlada_old_learn_mode-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Maui matl√•da (old)  Learn Mode',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.maui_matlada_old_learn_mode',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
